lib/DxcSupport/dxcapi.extval.cpp

```diff
@@ -4,39 +4,255 @@
 // WinIncludes must come before dxcapi.extval.h
 #include "dxc/Support/dxcapi.extval.h"
 
+std::vector<LPCWSTR> AddVDCompilationArg(UINT32 ArgCount, LPCWSTR *pArguments) {
+  UINT32 newArgCount = ArgCount + 1;
+  std::vector<LPCWSTR> newArgs;
+  newArgs.reserve(newArgCount);
+
+  // Copy existing args
+  for (unsigned int i = 0; i < ArgCount; ++i) {
+    newArgs.push_back(pArguments[i]);
+  }
+
+  // Add an extra argument
+  newArgs.push_back(L"-Vd");
+  return newArgs;
+}
+
 namespace dxc {
 
+// The purpose of the ExternalValidationHelper* wrappers is
+// to wrap the Compile call so that -Vd can be passed, disabling
+// internal validation. Then, we can allow the caller to handle
+// external validation. This wrapper is used when in an external
+// validation scenario, so there is no point in also running
+// the internal validator
+class ExternalValidationHelper3 : public IDxcCompiler3 {
+private:
+  CComPtr<IDxcCompiler3> m_pCompiler3;
+  CComPtr<IDxcValidator> m_pValidator;
+
+public:
+  HRESULT STDMETHODCALLTYPE Compile(const DxcBuffer *pSource,
+                                    LPCWSTR *pArguments, UINT32 argCount,
+                                    IDxcIncludeHandler *pIncludeHandler,
+                                    REFIID riid, LPVOID *ppResult) override {
+    // First, add -Vd to the compilation args, to disable the
+    // internal validator
+    std::vector<LPCWSTR> newArgs = AddVDCompilationArg(argCount, pArguments);
+    return m_pCompiler3->Compile(pSource, newArgs.data(), newArgs.size(),
+                                 pIncludeHandler, riid, ppResult);
+  }
+
+  HRESULT STDMETHODCALLTYPE Disassemble(const DxcBuffer *pObject, REFIID riid,
+                                        LPVOID *ppResult) override {
+    return m_pCompiler3->Disassemble(pObject, riid, ppResult);
+  }
+
+  HRESULT STDMETHODCALLTYPE QueryInterface(REFIID riid,
+                                           void **ppvObject) override {
+    return m_pCompiler3->QueryInterface(riid, ppvObject);
+  }
+
+  ULONG STDMETHODCALLTYPE AddRef() override { return m_pCompiler3.p->AddRef(); }
+
+  ULONG STDMETHODCALLTYPE Release() override {
+    return m_pCompiler3.p->Release();
+  }
```

Comment:
We can't just forward AddRef and Release like this.  Release() needs to delete `this` when the refcount reaches zero, which means that these wrapper classes will leak.

You'll to provide proper implementations of AddRef and Release.